File: Exercises/101-109 - Small Problems/easy_6/halvsies.rb
# Write a method that takes an Array as an argument, and returns two Arrays that
# contain the first half and second half of the original Array, respectively.
# If the original array contains an odd number of elements, the middle element should
# be placed in the first half Array.

# Examples:

# ```
# halvsies([1, 2, 3, 4]) == [[1, 2], [3, 4]]
# halvsies([1, 5, 2, 4, 3]) == [[1, 5, 2], [4, 3]]
# halvsies([5]) == [[5], []]
# halvsies([]) == [[], []]
# ```

def halvsies(arr)
  first_half = arr.slice(0, (arr.size / 2.0).ceil)
  second_half = arr.slice(first_half.size, arr.size - first_half.size)
  [first_half, second_half]
end

puts halvsies([1, 2, 3, 4]) == [[1, 2], [3, 4]]
puts halvsies([1, 5, 2, 4, 3]) == [[1, 5, 2], [4, 3]]
puts halvsies([5]) == [[5], []]
puts halvsies([]) == [[], []]

File: Exercises/210 - Small Problems/easy-4/halvsies.js
/*
Write a function that takes an Array as an argument, and returns an array that contains 2 elements, each of which is an Array. Put the first half of the original Array elements in the first element of the return value, and put the second half in the second element. If the original array contains an odd number of elements, place the middle element in the first half Array.

Examples:

halvsies([1, 2, 3, 4])    // [[1, 2], [3, 4]]
halvsies([1, 5, 2, 4, 3]) // [[1, 5, 2], [4, 3]]
halvsies([5])             // [[5], []]
halvsies([])              // [[], []]
*/

function halvsies(arr) {
  var half = Math.ceil(arr.length / 2);
  var firstHalf = arr.slice(0, half);
  var secondHalf = arr.slice(half);

  return [firstHalf, secondHalf];
}

console.log(halvsies([1, 2, 3, 4]));
console.log(halvsies([1, 5, 2, 4, 3]));
console.log(halvsies([5]));
console.log(halvsies([]));